doc/decisions/DR024-Versioning-traits-and-specifications-method.md

```diff
@@ -0,0 +1,365 @@
+
+def success_cb(idx, traits_data):
+    result[0] = traits_data
+
+def error_cb(idx, batchElementError):
+    if instanceOf(batchElementError, VersionNotSupportedError):
+      incompatible_with_manager_error()
+    ...
+
+self._manager.resolve(
+    [entity_reference], trait_set, ResolveAccess.kRead, context, success_cb, error_cb
+)
+```
+### Manager side
+```python
+# On the manager side, each trait gets unpacked (probably via a utility
+# method provided by traitgen or openassetio), and branching is done
+# on a trait by trait basis
+
+def resolve(
+        self,
+        entityReferences,
+        traitSet,
+        resolveAccess,
+        context,
+        hostSession,
+        successCallback,
+        errorCallback,
+    ):
+
+  # This function is imagined a pre-provided utility, possibly generated
+  # via traitgen, that would error in some way if the trait set does not
+  # contain the locatableContent trait.
+  locatableContent_version = locatableContent_version_from_trait_set(traitSet)
+
+  success_result = TraitsData()
+  if (locatableContent_version == 2):
+    trait = mediacreation.traits.content.LocatableContentTrait(success_result)
+    trait.setUrl(managed_filesystem_locations[ref.toString()])
+    successCallback(idx, success_result)
+  elif (locatableContent_version == 1):
+    trait = mediacreation.legacy.traits.content.LocatableContentTrait_v1(success_result)
+    trait.setLocation(managed_filesystem_locations[ref.toString()])
+    successCallback(idx, success_result)
+  else:
+    raise RuntimeError()
+
+```
+
+#### Pros
+- Both managers and hosts have the option to be backwards compatible.
+- More flexible, traits of different versions can be mixed in the same
+  session, which could allow hosts/managers to perform partial upgrades
+  in the case they are blocked on small specifics.
+- Hosts don't have to interrogate and declare which library versions
+  are in use, which helps especially in the case of second-order
+  dependencies.
+- Seems to fit with existing OpenAssetIO design, especially with
+  `BatchElementError`, as an incompatible version wouldn't necessarily
+  need to error the whole batch.
+- No new API, core `openassetio` library remains version unaware.
+- Cannot be ignored, unlike other "layered on top" mechanisms.
+
+#### Cons
+- No "primary" version across the session, increasing implementation
+  complexity.
+- Introduces "Try and see" negotiation on the host side, although
+  introspection behaviour to negate this could be introduced at a later
+  date, similar to other introspection mechanisms. (See Option 3 -
+  Extended)
+- Would likely require some utilities in `traitgen`, otherwise, due to
+  each new version effectively being a brand new trait, constructing the
+  correct trait view would become a tedious dance.
+- EntityTraits becomes more complex, as we either have to declare it
+  version agnostic somehow, or managers would need to return the correct
+  versions/entire swathes of versions.
```

Comment:
[consider][minor] Is `entityTraits` really a special case? In the simplest formulation, its the mirror on the host side of what the manager must do, i.e. branch on every trait/specification its interested in.

I.e. re. "managers would need to return the correct versions/entire swathes of versions" - you could apply the same argument to `resolve` - the host could resolve swathes of versions.

It's back to that question of "who's in charge here".

```suggestion
- Hosts and managers must detect and branch on every supported trait version and/or communicate all their supported versions in an API request (e.g. host->manager via `resolve`, manager->host via `entityTraits`).
```

Reply:
Resolve was a given in my mind, the reason to call out `entityTraits` to my mind is this is a specific drawback of an in-data solution.`entityTraits` is asking like "what is this entity, what does it do?", and when asking that question, the version information seems _almost_ redundant. The in-data approach forces versioning in a first class concern of every method, as you say, but that uniquely influences `entityTraits` imo.

Reply:
I think the “who’s in charge here” thing is key. 
If it’s this becomes a democratic negotiatiation to a happy place, then the code is in charge somehow (and that feels confusing as it’s not clear from an API perspective who ultimately makes a decision). An alternative way to look at it, which is what you get in file world, (which is more similar than perhaps thought, if you abstract the transport layer as we do and desire caching middleware) where the **data** is in charge. The host will provide vX of data, the manager may provide vY of data, its at the point that specific interpretation is needed where conversion lies, not the transport layer. So either end can be “simple” as it it’s just becoming part of the transport. Rather than reading both parties to be active, all the time, which creates the “swathes” problem. `entityTraits` then is simply either the actual version of the data - in the simple BAL case, or “the preferred/only version of the code used by the manager” in dynamic translation cases.

A key thought experiment for me on this all would be “what does BAL, simpleResolver and a the generic relationship/entity browser look like with these two approaches. And critically how does their use/functionality change.

Re versioning being redundant in entity traits - if different versions of traits are incompatible the is versioning actually very important in the entity traits method? As an entity having the qualities of one version or another may be very different! Eg adding a property to a trait would mean entities with the older version are inherently not going to have that data - which may be required (I think we suggested leaning towards small traits with only required properties right?)

Reply:
There's then perhaps a separate Con that "unversioniong" of some methods may be required.  

You could conceivably apply the argument to `resolve` - if you assume that the host can branch on the `TraitsData` traits returned from `resolve`, then the input to `resolve` could be unversioned trait IDs.

I think my argument that `entityTraits` isn't necessarily a special case holds. It's perhaps more susceptible to a rethink of whether it should be unversioned than other methods, but not exclusively.

Reply:
Yeah. I guess one nice thing with the in-data option is it isn't an API signature change just a calling convention change, so the decisions around that are relatively easy to iterate. It might even be that both make sense. Eg a host resolves a specific version if that's all it supports, or leaves it unversioned in the utopian 'auto up/down grade traitgen world'

Reply:
This got simplified in general, there's a whole section about this so the con can just link now : https://github.com/OpenAssetIO/OpenAssetIO/pull/1267/commits/066101d9e574e43e101aea7e2590ddf0527b2c3d